modules/tensor_mechanics/include/materials/crystal_plasticity/CrystalPlasticityStressUpdateBase.h

```diff
@@ -0,0 +1,226 @@
+   * and to normalize the vectors. This method is abstracted to allow for reuse
+   * in inheriting classes with multiple plane normal and direction vector pairs.
+   */
+  void getPlaneNormalAndDirectionVectors(const FileName & vector_file_name,
+                                         const unsigned int & number_dislocation_systems,
+                                         DenseVector<Real> & plane_normal_vector,
+                                         DenseVector<Real> & direction_vector,
+                                         bool & orthonormal_error);
+
+  /**
+   * A helper method to sort the slip systems of a crystal into cross slip families based
+   * on common slip directions.  This method determines if slip directions are parallel,
+   * and stores the index of the slip systems from getSlipSystems (the same index is used
+   * for the applied shear stress and the internal state variables) in a vector of vectors,
+   * where the outer vector separates the individual slip system families and the inner vector
+   * stories the indices of the slip systems within a single family.  This vector of vectors
+   * can then be used in the inheriting classes to calculate cross slip interactions.
+   * The values of number_cross_slip_directions and number_cross_slip_planes must be set to
+   * use this sorting method.
+   */
+  void sortCrossSlipFamilies();
+
+  /**
+   * A helper method for inherting classes to identify to which cross slip family vector
+   * a particular slip system index belongs after the slip systems have been sorted. Returns
+   * the integer value of the identified cross slip system family for the outer vector
+   * created in sortCrossSlipFamilies.
+   */
+  unsigned int identifyCrossSlipFamily(const unsigned int index);
+
+  /**
+   * This virtual method is called to set the constitutive internal state variables
+   * current value and the previous substep value to the old property value for
+   * the start of the stress convergence while loop.
+   */
+  virtual void setInitialConstitutiveVariableValues() {}
+
+  /**
+   * This virtual method is called to set the current constitutive internal state
+   * variable value to that of the previous substep at the beginning of the next
+   * substep increment. In cases where only one substep is taken (or when the first)
+   * substep is taken, this method sets the current constitutive internal state
+   * variable value to the old value.
+   */
+  virtual void setSubstepConstitutiveVariableValues() {}
+
+  /**
+   * Stores the current value of the constitutive internal state variables into
+   * a separate material property in case substepping is required, once the
+   * constitutive variables have passed convergence tolerances. This separate
+   * material property is used as the previous substep value in the associated
+   * setSubstepConstitutiveVariableValues method in the next substep (if taken).
+   */
+  virtual void updateSubstepConstitutiveVariableValues() {}
```

Comment:
Suggestion for better function names? @bwspenc @dschwen

Reply:
I love it already

Reply:
...said Ben _never_  ;-)